Mathlib/Combinatorics/SimpleGraph/Walk.lean

```diff
@@ -765,61 +770,81 @@ lemma notNilRec_cons {motive : {u w : V} → (p : G.Walk u w) → ¬ p.Nil → S
     motive (q.cons h) Walk.not_nil_cons) (h' : G.Adj u v) (q' : G.Walk v w) :
     @Walk.notNilRec _ _ _ _ _ cons _ _ = cons h' q' := by rfl
 
-/-- The second vertex along a non-nil walk. -/
-def sndOfNotNil (p : G.Walk v w) (hp : ¬ p.Nil) : V :=
-  p.notNilRec (@fun _ u _ _ _ => u) hp
-
-@[simp] lemma adj_sndOfNotNil {p : G.Walk v w} (hp : ¬ p.Nil) :
-    G.Adj v (p.sndOfNotNil hp) :=
-  p.notNilRec (fun h _ => h) hp
+@[simp] lemma adj_getVert_one {p : G.Walk v w} (hp : ¬ p.Nil) :
+    G.Adj v (p.getVert 1) := by
+  have := adj_getVert_succ p (by
+    rw [SimpleGraph.Walk.nil_iff_length_eq] at hp
+    push_neg at hp
+    exact Nat.zero_lt_of_ne_zero hp)
+  rwa [getVert_zero] at this
```

Comment:
This proof really should be
```suggestion
  simpa using adj_getVert_succ p (by simpa)
```
Can you add simp lemmas until something of the above works?

Reply:
`simpa using adj_getVert_succ p (by simpa : 0 < p.length)` works except for the second simpa, but simping "not nil" to "0 < p.length" seems pretty impactful, are you sure we want to be doing that?

Reply:
I managed to get it to this:

```
@[simp] lemma adj_getVert_one {p : G.Walk v w} (hp : ¬ p.Nil) :
    G.Adj v (p.getVert 1) := by
  have := adj_getVert_succ p (by simpa [not_nil_iff_lt_length] using hp)
  simp only [getVert_zero, zero_add] at this
  exact this
```

which works

but

```
@[simp] lemma adj_getVert_one {p : G.Walk v w} (hp : ¬ p.Nil) :
    G.Adj v (p.getVert 1) := by
  simpa [getVert_zero, zero_add] using adj_getVert_succ p (by simpa [not_nil_iff_lt_length] using hp)
```
errors with

```
type mismatch
  h✝
has type
  G.Adj (p.getVert 0) (p.getVert (0 + 1)) : Prop
but is expected to have type
  G.Adj v (p.getVert 1) : Prop
```

Any clue what's going on here?

Reply:
Hmm... no clue, no. If you manage to minimise it, try posting it on #mathlib4 or #lean4 (depending on whether you got rid of the mathlib dependencies or not)

Reply:
Probably some metavariable instantiating too late